src/main/java/io/cryostat/discovery/DiscoveryNode.java

```diff
@@ -80,6 +81,30 @@ public class DiscoveryNode extends PanacheEntity {
     @Nullable
     public List<DiscoveryNode> children = new ArrayList<>();
 
+    @JsonView(Views.Nested.class)
+    @Transient
+    public List<DiscoveryNode> descendantTargets;
+
+    public List<DiscoveryNode> getDescendantTargets() {
+        if (descendantTargets == null) {
+            descendantTargets = fetchDescendantTargets(this);
```

Comment:
Couple of things here:
1. This isn't thread-safe - if two different threads calls `getDescendantTargets()` at the same time, then both of them will end up executing the actual fetch method, both allocating a new ArrayList, and it will just come down to timing/scheduling which ones' result actually updates the reference last and sets the value on the heap which others will see in the future
2. This list should probably be externally immutable, ie. I shouldn't be allowed to do something like `node.getDescendantTargets().clear()` from some other class.
3. The overall discovery tree is mutable, ie nodes come and go, and targets come and go. If the node we're looking at is the Universe node for example, then this method should return all of the discovered targets whenever it is called. However, because of this caching strategy, the list of descendants will only be calculated once ever (other than multithreading bugs in point 1). From that point on, no matter what discovery changes occur, this method will always return the same result.

Reply:
It looks like this only gets called from within the tests as well. If we're going to keep this method, then there is some refactoring to be done:

https://github.com/cryostatio/cryostat/blob/7083a7f3cf8450880ff7edd45be76d3bc51f88fd/src/main/java/io/cryostat/graphql/RootNode.java#L48